Changes in module 'nf-core/gatk4/splitncigarreads'
--- modules/nf-core/gatk4/splitncigarreads/main.nf
+++ modules/nf-core/gatk4/splitncigarreads/main.nf
@@ -4,17 +4,20 @@
 
     conda "bioconda::gatk4=4.4.0.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/gatk4:4.4.0.0--py36hdfd78af_0':
-        'biocontainers/gatk4:4.4.0.0--py36hdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/mulled-v2-d9e7bad0f7fbc8f4458d5c3ab7ffaaf0235b59fb:f857e2d6cc88d35580d01cf39e0959a68b83c1d9-0':
+        'biocontainers/mulled-v2-d9e7bad0f7fbc8f4458d5c3ab7ffaaf0235b59fb:f857e2d6cc88d35580d01cf39e0959a68b83c1d9-0' }"
 
     input:
-    tuple val(meta), path(bam), path(bai), path(intervals)
+    tuple val(meta), path(input), path(input_index), path(intervals)
     path  fasta
     path  fai
     path  dict
 
     output:
-    tuple val(meta), path('*.bam'), emit: bam
+    tuple val(meta), path("*cram"),     emit: cram,  optional: true
+    tuple val(meta), path("*bam"),      emit: bam,   optional: true
+    tuple val(meta), path("*.crai"),    emit: crai,  optional: true
+    tuple val(meta), path("*.bai"),     emit: bai,   optional: true
     path  "versions.yml"          , emit: versions
 
     when:
@@ -22,7 +25,11 @@
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    prefix = task.ext.prefix ?: "${meta.id}.bam"
+    // If the extension is CRAM, then change it to BAM
+    prefix_bam = prefix.tokenize('.')[-1] == 'cram' ? "${prefix.substring(0, prefix.lastIndexOf('.'))}.bam" : prefix
+
     def interval_command = intervals ? "--intervals $intervals" : ""
 
     def avail_mem = 3072
@@ -33,16 +40,25 @@
     }
     """
     gatk --java-options "-Xmx${avail_mem}M" SplitNCigarReads \\
-        --input $bam \\
-        --output ${prefix}.bam \\
+        --input $input \\
+        --output ${prefix_bam} \\
         --reference $fasta \\
         $interval_command \\
         --tmp-dir . \\
         $args
 
+    # If cram files are wished as output, the run samtools for conversion
+    if [[ ${prefix} == *.cram ]]; then
+        samtools view -Ch -T ${fasta} -o ${prefix} ${prefix_bam}
+        rm ${prefix_bam}
+        samtools index ${prefix}
+    fi
+
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
     END_VERSIONS
     """
 }

************************************************************
