Changes in module 'nf-core/picard/filtersamreads'
--- modules/nf-core/picard/filtersamreads/main.nf
+++ modules/nf-core/picard/filtersamreads/main.nf
@@ -9,10 +9,12 @@
 
     input:
     tuple val(meta), path(bam), path(readlist)
+    path fasta
     val filter
 
     output:
     tuple val(meta), path("*.bam"), emit: bam
+    tuple val(meta), path("*.bai"), emit: bai
     path "versions.yml"           , emit: versions
 
     when:
@@ -35,6 +37,7 @@
         picard \\
             FilterSamReads \\
             -Xmx${avail_mem}M \\
+            -R $fasta \\
             --INPUT $bam \\
             --OUTPUT ${prefix}.bam \\
             --FILTER $filter \\
@@ -50,6 +53,7 @@
         picard \\
             FilterSamReads \\
             -Xmx${avail_mem}M \\
+            -R $fasta \\
             --INPUT $bam \\
             --OUTPUT ${prefix}.bam \\
             --FILTER $filter \\

************************************************************
